cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/IndexedTable.java

```diff
@@ -532,22 +557,20 @@ public IndexScanner(Scanner baseScanner, byte[] rowKeyPrefix) {
     @Override
     public Row next() {
       // TODO: retrieve results in batches to minimize RPC overhead (requires multi-get support in table)
-      Row dataRow = null;
       // keep going until we hit a non-null, non-empty data row, or we exhaust the index
-      while (dataRow == null) {
-        Row indexRow = baseScanner.next();
-        if (indexRow == null) {
-          // end of index
-          return null;
-        }
+      Row indexRow;
+      while ((indexRow = baseScanner.next()) != null) {
         byte[] rowkey = indexRow.get(IDX_COL);
         // verify that datarow matches the expected row key to avoid issues with column name or value
         // containing the delimiter used
-        if (rowkey != null && Bytes.equals(indexRow.getRow(), Bytes.add(rowKeyPrefix, rowkey))) {
-          dataRow = table.get(rowkey);
+        if (rowkey != null) {
```

Comment:
this is incorrect. Suppose I have two indexed fields: a and a\0b. A row x has a=b\0c, and a row y has a\0b=c. When scanning on column a with the range b..c, I should only see row x. But I will see both rows. 

Actually this was incorrect in the existing IndexedTable. I think after fetching the actual row, we have to double-check that it is a correct result.